operator/cmd/allocator_test.go

```diff
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/cilium/ipam/cidrset"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	k8s_testing "k8s.io/client-go/testing"
+
+	"github.com/cilium/cilium/pkg/ipam/allocator/podcidr"
+	"github.com/cilium/cilium/pkg/ipam/types"
+	cilium_api_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	cilium_fake "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/fake"
+)
+
+func TestPodCIDRAllocatorOverlap(t *testing.T) {
+	// Create a new CIDR allocator
+
+	_, cidr, err := net.ParseCIDR("10.129.0.0/16")
+	if err != nil {
+		panic(err)
+	}
+
+	set, err := cidrset.NewCIDRSet(cidr, 24)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create a mock APIServer client where we have 2 existing nodes, one with a PodCIDR and one without.
+	// When List'ed from the client, first node-a is returned then node-b
+
+	fakeClient := cilium_fake.NewSimpleClientset(&cilium_api_v2.CiliumNode{
+		ObjectMeta: v1.ObjectMeta{
+			Name: "node-a",
+		},
+		Spec: cilium_api_v2.NodeSpec{
+			IPAM: types.IPAMSpec{
+				PodCIDRs: []string{},
+			},
+		},
+	}, &cilium_api_v2.CiliumNode{
+		ObjectMeta: v1.ObjectMeta{
+			Name: "node-b",
+		},
+		Spec: cilium_api_v2.NodeSpec{
+			IPAM: types.IPAMSpec{
+				PodCIDRs: []string{
+					"10.129.0.0/24",
+				},
+			},
+		},
+	})
+
+	// Sync channel to know when the APIServer state is stable.
+	updated := make(chan struct{})
+	var timer *time.Timer
+
+	// Add a reactor to the fake client, the callback is triggered for each Update to the APIServer.
+	poolReactor := &k8s_testing.SimpleReactor{
+		Verb:     "update",
+		Resource: "ciliumnodes",
+		Reaction: func(action k8s_testing.Action) (handled bool, ret runtime.Object, err error) {
+			// Close `updated` 1 second after the last update
```

Comment:
Why?

Reply:
Later on in the code we block on `updated` the idea here is that we want to know when we are done processing. We don't really have an indicator other than updates on the mocked API server. This detects the burst of updates by closing the channel once we have seen at least one update and it has been 1 second since the last update we have seen.

Reply:
This was a hint to leave a comment in the code since it's not obvious :)